File: lib/services/rental_service.rb
require 'yaml'
require_relative '../models/item'
require_relative '../models/rental'

module Services
  class RentalService
    attr_reader :items, :rentals

    def initialize
      @items = load_items
      @rentals = []
    end

    def load_items
      items_data = YAML.load_file('data/items.yml')
      items_data.map.with_index(1) { |name, index| Models::Item.new(index, name) }
    end

    def list_items
      @items.each { |item| puts "#{item.id}. #{item.name}" }
    end

    def find_item_by_id(id)
      @items.find { |item| item.id == id }
    end

    def item_available?(item, start_date, end_date)
      @rentals.none? { |r| r.item.id == item.id && r.overlaps?(start_date, end_date) }
    end

    def add_rental(rental)
      @rentals << rental
    end

    def active_rentals
      @rentals.select(&:active?)
    end
  end
end

File: lib/services/rental_service.dart
import '../models/item.dart';
import '../models/rental.dart';

class RentalService {
  final List<Item> items = [];
  final List<Rental> rentals = [];

  RentalService() {
    _populateItems();
  }

  void _populateItems() {
    items.addAll([
      Item(1, 'Camera'),
      Item(2, 'Bicycle'),
      Item(3, 'Laptop'),
      Item(4, 'Projector'),
    ]);
  }

  List<Item> getAvailableItems() {
    return items;
  }

  bool isItemAvailable(Item item, DateTime start, DateTime end) {
    for (var rental in rentals) {
      if (rental.item.id == item.id && rental.overlaps(start, end)) {
        return false;
      }
    }
    return true;
  }

  Rental? getConflictingRental(Item item, DateTime start, DateTime end) {
    for (var rental in rentals) {
      if (rental.item.id == item.id && rental.overlaps(start, end)) {
        return rental;
      }
    }
    return null;
  }

  void addRental(Rental rental) {
    rentals.add(rental);
  }

  List<Rental> getActiveRentals() {
    return rentals.where((r) => r.isActive()).toList();
  }
}
